components/sync15/src/changeset.rs

```diff
@@ -61,18 +61,28 @@ impl OutgoingChangeset {
 impl IncomingChangeset {
     pub fn fetch(
         client: &Sync15StorageClient,
-        state: &GlobalState,
+        state: &mut CollectionState,
         collection: String,
         collection_request: &CollectionRequest,
     ) -> Result<IncomingChangeset> {
-        let records = client.get_encrypted_records(collection_request)?;
-        let timestamp = state.last_modified_or_zero(&collection);
+        let (records, timestamp) = match client.get_encrypted_records(collection_request)? {
+            Sync15ClientResponse::Success {
+                record,
+                last_modified,
+                ..
+            } => (record, last_modified),
+            other => return Err(other.to_storage_error().into()),
+        };
+        // xxx - duplication below of `timestamp` smells wrong
+        state.last_modified = timestamp;
         let mut result = IncomingChangeset::new(collection, timestamp);
         result.changes.reserve(records.len());
-        let key = state.key_for_collection(&result.collection)?;
         for record in records {
-            // TODO: if we see a HMAC error, may need to update crypto/keys?
-            let decrypted = record.decrypt(&key)?;
+            // if we see a HMAC error, we've made an explicit decision to
+            // NOT handle it here, but restart the global state machine.
+            // TODO: However, it's not clear how that global state machine
+            // knows this.
```

Comment:
Yeah, I think this will work fine. With this branch, we will probably attempt to sync subsequent engines too, but they'll also fail with the same error - but these failures will cause us to drop our memory state, so the next sync will (theoretically ;) do the right thing.

The only reason I haven't done that in this patch is that it's not immediately clear how we could classify errors into "abort this sync entirely" vs "some transient error specific to the collection, so it's possible other collections will work" - it seems better to err on the side of "just try the other engines" because no harm should come from them failing in the same way the earlier engine did.

Reply:
(oh, and I meant to mention - this is the reason for the under-utilized Sync15ClientResponse() - the `Unauthorized` and `ServerError` variants in particular could help drive that decision.)